app/services/distribution_service.rb

```diff
@@ -0,0 +1,55 @@
+class DistributionService
+  attr_reader :error
+
+  def perform_distribution_service(&block)
+    distribution.transaction do
+      yield block
+    end
+  rescue ActiveRecord::RecordNotFound => e
+    Rails.logger.error "[!] #{self.class.name} failed to destroy distribution #{distribution_id} because it does not exist"
+    set_error(e)
+  rescue Errors::InsufficientAllotment => e
+    distribution.line_items.assign_insufficiency_errors(e.insufficient_items)
+    Rails.logger.error "[!] #{self.class.name} failed because of Insufficient Allotment #{distribution_organization.short_name}: #{distribution.errors.full_messages} [#{e.message}]"
+    set_error(e)
+  rescue StandardError => e
+    Rails.logger.error "[!] #{self.class.name} failed to destroy distribution for #{distribution_organization.short_name}: #{distribution.errors.full_messages} [#{e.inspect}]"
+    set_error(e)
+  ensure
+    return self
+  end
+
+  def success?
+    error.nil?
+  end
+
+  def distribution
+    # Return distribution if it has already been defined
+    return @distribution if @distribution
+
+    # Otherwise try to get this value with possibly
+    # provided distribution_id from initialize
+    if @distribution_id.present?
+      @distribution = Distribution.find(@distribution_id)
+    end
+  end
+
+  def distribution_id
+    return @distribution_id if @distribution_id
+
+    if distribution.present?
+      @distribution_id = distribution.id
+    end
+  end
```

Comment:
These two functions make me a little nervous if they were to be aclled from outside of the error messages - could we make them private maybe?

Reply:
It looks like this method is only used in one place -- in an error message. I bet we could remove this and just reference it via the `:distribution` method instead. We shouldn't have a distribution id without there being a pre-existing distribution (and if we're looking for one that doesn't exist, that could just raise `ActiveRecord::RecordNotFound`?)